model/jpa/src/main/java/org/keycloak/grant/jpa/impl/GrantServiceImpl.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.grant.jpa.impl;
+
+import org.keycloak.models.*;
+import org.keycloak.grant.jpa.UserGrantEntity;
+import org.keycloak.models.jpa.entities.UserConsentEntity;
+import org.keycloak.storage.StorageId;
+
+import javax.persistence.EntityManager;
+import javax.persistence.LockModeType;
+import javax.persistence.TypedQuery;
+import java.util.List;
+
+public class GrantServiceImpl implements GrantService {
+
+    private final KeycloakSession session;
+    protected EntityManager em;
+
+    public GrantServiceImpl(KeycloakSession session, EntityManager em) {
+        this.session = session;
+        this.em = em;
+        if (getRealm() == null) {
+            throw new IllegalStateException("The service cannot accept a session without a realm in its context.");
+        }
+    }
+
+    protected RealmModel getRealm() {
+        return session.getContext().getRealm();
+    }
+
+
+    @Override
+    public boolean revokeGrantByGrantId(RealmModel realm, String grandId, String userId) {
```

Comment:
@guymoyo AFAIK, there is no trivial solution. If you decide such that a newly introduced Grant corresponds with an Authenticated Client Session , you can take the same way as in a refresh token's token revocation. Please refer to `TokenRevocationEndpoint`. 
When receiving a refresh token on this endpoint, keycloak revokes corresponding Authenticated Client Session which leads to revoking a refresh token.

The way depends on the determination of the relationships between Grant, User Session, Authenticated Client Session, Access Token and Refresh Token. The simple way is to define such that 1 Grant corresponds with 1 Authenticated Client Session.

Keycloak does not manage access tokens and refresh tokens one by one. Instead of that, keycloak manages User Session, Authenticated Client Session and its latest updated refresh token. keycloak manages a consent per (user, client) pair.

Reply:
I see, a Grant with a corresponding Authenticated Client Session, we can revoke all the Refresh Token. But the Access Token will still be valid (although for a defined limited frame of time).

The only way I see now for revoking Access Token is to keep the AccessTokenId along with the grant.